arwen/host/asyncCall.go

```diff
@@ -0,0 +1,150 @@
+package host
+
+import (
+	"math/big"
+
+	arwen "github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/wasmer"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func (host *vmHost) handleAsyncCallBreakpoint(result wasmer.Value) error {
+	runtime := host.Runtime()
+	runtime.SetRuntimeBreakpointValue(arwen.BreakpointNone)
+
+	// TODO also determine whether caller and callee are in the same Shard (based
+	// on address?), by account addresses - this would make the empty SC code an
+	// unrecoverable error, so returning nil here will not be appropriate anymore.
+	if !host.canExecuteSynchronously() {
+		return nil
+	}
+
+	// Start calling the destination SC, synchronously.
+	destinationCallInput, err := host.createDestinationContractCallInput()
+	if err != nil {
+		return err
+	}
+
+	destinationVMOutput, err := host.ExecuteOnDestContext(destinationCallInput)
+	if err != nil {
+		return err
+	}
+
+	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput)
+	if err != nil {
+		return err
+	}
+
+	_, err = host.ExecuteOnDestContext(callbackCallInput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (host *vmHost) canExecuteSynchronously() bool {
+	runtime := host.Runtime()
```

Comment:
take destination address and verify if it is in current shard. We can make a blockchain hook for it. canExecuteSynchronously will do only that.

Reply:
Added TODO for now, noted for next development steps.